PlatformUI BDD core implementations

miguelcleverti opened the pull request:
# PlatformUI BDD base
Base implementations for platformUI BDD
Needed to run the PlatformUI bundle scenarios

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     */
+    protected function clickLogo()
+    {
+        $selector = 'img[alt="Logo eZ"]';
```
to make it a bit more robust, I would use `.ez-logo img` as a selector instead (the `alt` value is more a data than a markup thing)

miguelcleverti replied:
yes, you are right I didn't spend much time on the logo sentence because it was not a priority at the time but I will fix it.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+    {
+        $this->shouldBeLoggedIn = true;
+
+        $jsCode = <<<JSCODE
```
eh what is that ? if it works, you are lucky, since the use callback function can be called asynchronously.
if you just want to check if an element is in the page, you don't need YUI at all, something like:

```js
return (document.querySelector('.ez-loginform') === null);
```

should do the trick. But that does not really check that you are logged but more that the login form disappeared.

miguelcleverti replied:
Yes it is more simpler this way. This check was implemented because there were problems with platformUI losing the log in, this was a quick way for we to check that everything was going okay. I will take a note to improve this.

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $text
+     */
+    protected function clickMenuZone( $text )
```
to fit the domain, this method should be called `clickNavigationZone`

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     */
+    protected function clickMenuLink( $text )
+    {
+        $this->clickElementsByText( $text, '.ez-link a[href]' );
```
the `ez-link` links is gone for quite some time now, so I guess this method can be removed

miguelcleverti replied:
Yes, I inteded to remove this but forgot

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $text
+     */
+    protected function clickSubMenuLink( $text )
```
to fit with the domain, this should be called `clickNavigationItem` (as the class suggests)

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $text
+     */
+    protected function clickSideMenuLink( $text )
```
same here, `clickActionBarAction` or something like that, *side menu* does not refer to anything in PlatformUI.

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     */
+    protected function clickSideMenuLink( $text )
+    {
+        $this->clickElementsByText( $text, '.ez-action p.action-label' );
```
not sure you really need to be that specific, `.ez-action` should be enough

miguelcleverti replied:
This is actually a remnant of an earlier implementation that needed to be more specific, after the improvement this missed the clean up

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     */
+    protected function clickContentType( $text )
+    {
+        $this->clickElementsByText( $text, '.ez-selection-filter-item' );
```
this is not specific enough, the selection filter can be used in others contexts `.ez-contenttypeselector-types .ez-selection-filter-item` would be ok I guess

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $name
+     */
+    protected function clickElementsByText( $text, $selector )
```
since you are clicking on one element, it should be called `clickElementByText`

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $name
+     */
+    protected function checksElementsByText( $text, $selector )
```
`checksElementByText`

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+    protected function attachFilePrepare( $path, $selector )
+    {
+        $jsArgs = JsHelper::generateFuncArgs( $selector );
+        $jsCode = "return BDD.changeCssDisplay( $jsArgs );";
```
hum that's a bit weird to have to unhide the file input to fill it, is that really needed ?

miguelcleverti replied:
Yes, I struggled with this for some time and the only way I found was to unhide the file input, the problem is that the webdriver can't attach a file to an hidden input.

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+
+            if ( is_file( $fullPath ) )
+            {
+                $fileInput = 'input[type="file"]';
```
if there's many file inputs, you might fill the wrong one.

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+                {
+                    throw new Exception( "File input is not found" );
+                }
+            $field->attachFile( $fullPath );
```
nitpick: wrong indentation level

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * Finds an HTML element by class and the text value and return the value of the sibling
```
wrong php doc

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @param string $name
+     */
+    protected function checksImagePresent( $text, $selector )
```
what is the `$text` variable for ?

miguelcleverti replied:
The text is needed to find the image, normally associated with a text label, example:
![image](https://cloud.githubusercontent.com/assets/8983122/6462859/c110484e-c1a4-11e4-83d8-f31732b585c9.png)

$text in red

bdunogier replied:
Then it should be called "label", or "imageLabel", should it not ?

miguelcleverti replied:
I will change it, I guess $label is more explicit.

bdunogier replied:
Well, you did find the right answer by yourself. If you can use a word to describe it in the sentence, then it's usually the right word. And if the word you have doesn't make sense when used directly in a sentence, then it is not right.

miguelcleverti replied:
makes sense :-)

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * Finds an HTML element by class and the text value and return the value of the sibling
```
wrong php doc

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * Get the current window Node Id
```
*current window Node Id* ? this does not mean anything, I guess this is returning the **location** id that is currently displayed but this only works when displaying a location not in the whole PlatformUI so this should probably not be in a `CommonActions` trait.

miguelcleverti replied:
Yes, this is a workaround because so far there is no information on platformUI about the content no ids, no type, etc. This is a supposed temporary solution for some test cases.

dpobel commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @return string
+     */
+    protected function getThisContentId()
```
`getLocationId`

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+     */
+    public function iUploadTheFile( $path )
+    {
+        $this->attachFilePrepare( $path, '#ez-field--file' );
```
this id is weird because it might not be unique across the page, so I guess there's a problem in the generated id in the page which should be something like `#ez-field-42-file` where 42 is the content id.

miguelcleverti replied:
I missed this, it is suposed to use the class tag, but while I was testing I used the id and forgot to change.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+     */
+    public function iUploadTheImage( $path )
+    {
+        $this->attachFilePrepare( $path, '#ez-field--image' );
```
same here

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+        //And I fill in "Password" with "publish"
+        $this->fillFieldWithValue( 'Password', $password );
+        //And I click on the "Login" button
+        $this->iClickAtButton( 'Login' );
```
`iClickOnButton( 'Login' );`

miguelcleverti replied:
The iClickAtButton funtion belong to BehatBundle browser context, but I can change it.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+    public function contentExists( $contentName, $contentType )
+    {
+        $contentId = $this->getThisContentId();
+        $content = $this->getContentManager()->loadContentWithLocationId( $contentId );
```
`loadContentWithLocationId( $contentId )` this is really confusing (same issue as in https://github.com/ezsystems/PlatformUIBundle/pull/181/files#r25683698)

miguelcleverti replied:
Like I said before, so far the only way I found to test some cases was uses the window hash to retrieve the location id, the variable name is wrong it should be $locationId, in the future this should probably be replace by a better method

bdunogier replied:
All @dpobel is referring to here is the naming, imho. And yes, it should be changed :-)

joaoinacio commented on Features/Lib/PlatformUI.js:
```diff
@@ -0,0 +1,422 @@
+    {
+        this.clickElement(
+                this.getElementByText(text, selector)
+                );
```
CS

dpobel replied:
on the CS thing, the whole file is wrong, the CS for JavaScript are different. For instance, this method should be written like:
```js
this.clickElementByText = function(text, selector) {
    this.clickElement(this.getElementByText(text, selector));
};
```
I guess there's enough JS example in this bundle to get it :-)

But, I'm a bit surprise to see such file here, shouldn't the BDD framework provide such helpers ?

pcardiga replied:
@andrerom: I agree. There might exist some information that might be included in the UI that would easily the tests and we could sync on that.

miguelcleverti replied:
BDD framework provides helpers for normal webpages, with AJAX and a full JS webpage things get a little more complicated. The JS was the only way we found to access certain elements and interact with them.

andrerom commented:
@miguelcleverti @pcardiga @dpobel Seems we might need to do a meeting to sync on approach, cause ideally the tests should not have to workaround test issues caued by how we have implemented the UI.

This is not a legacy UI, we can (and should) adjust how we generate the UI to simplify testing ;)

miguelcleverti commented:
@andrerom I agree, if the UI is generated with BDD in mind it will improve the overall quality of the tests and the UI itself

andrerom commented:
@miguelcleverti could you list up here the misc issues you encountered so far? (like not being able to use file upload because of hidden upload field* and so on).

_* A question that will pop up on this is if the tests really test the UI if it does not test the *faked* upload UI the user would work with, so might be we should make sure *that* is more testable somehow_.

miguelcleverti commented:
@andrerom 
- mentioned uplaod UI;
- little information about content (type, ids, etc);
- hidden elements cause wrong results with BDD base framework(behat), bypassed with JS;
- the Image API does not seem to have a loading tag;
- fields validation trigger could be improved for easier testing;
- application is loading tag are good overall but a global one("something is loading") would be even better;
- HTML generated structure could be greatly improved, following a simple consistent logic making the association between elements more explicit(most important IMO);
-Sahi can't publish contents, problems with cookies.

These are some of the main problems I have found. If we can improve this I believe it would benefit the UI. Also none of these are functionality related so it should be no problem.

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+<?php
+
+/**
+ * File containing the main context class for PlatformUI.
```
"This file is part of the PlatformUI package."

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,327 @@
+use Behat\Gherkin\Node\TableNode;
+use PHPUnit_Framework_Assert as Assertion;
+
+class PlatformUI extends Context
```
Please add a summary of what the class does, a couple lines (it looks like an important one).

bdunogier commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,363 @@
+                $jsFile = $this->execJavascript( $jsCode );
+            }
+        }
+        else throw new Exception( "File is not found at the given location" );
```
What file ? What location ? Try to think of what you'd like to be given as info when you encounter this exception in a month.

While I'm at it, CS:

```php
}
else
{
    throw...
}
```

miguelcleverti commented:
I fixed most of the issues commented here and improved the documentations in geral, also I added a doc folder with some relevant documents on installing/using the PlatformUI BDD

dpobel commented on Features/doc/README.md:
```diff
@@ -0,0 +1,28 @@
+* [PlatformUI bundle](https://github.com/ezsystems/PlatformUIBundle) is installed and configured properly.
+
+## Installation:
+* Update behat.yml, on the installation root, with PlatformUI parameters( check example behat.yml provided );
```
instead of documenting that, we should directly do it in the ezplatform's behat.yml.

dpobel replied:
or wouldn't it be possible to have a behat.yml in PlatformUIBundle that is just imported in the ezplatform's behat.yml ?

miguelcleverti commented:
Yes I also think it's better to add a platformUI only behat.yml but I think that some documentation should still be present since there are some parameters dependent of the enviroment. So it should be something like "Update the behat.yml on platformUI directory" and the example behat.yml could be deleted.

dpobel commented:
Documentation rarely hurts :-) That said, my point is more that the test should be directly ready to be run without any changes (except maybe the base_url). This would also ease the integration with Travis or any other CI platform.

Instead of adding lots of comments (still I will add some), I tried to experiment different things to remove most of PlatformUI.js which is more a hack than anything and sometimes does wrong/weird things that tends to fail from time to time. The result can be seen in https://github.com/ezsystems/PlatformUIBundle/commit/524c8b1f95de3ce8f78129a91104b9763d3193a2. This is an incomplete but I hope it gives you path to go:

1. `waitForJs` should not be used as is. As we discussed, it waits for way to many things. So I added a `waitForInitialApplicationLoading` and `waitForApplicationPageLoading` to wait for the app initial loading and to wait for the *page* loading. As, you can see, this does not rely at all on any custom JavaScript code but only on what [Mink provides to parse the HTML code of the page](http://mink.behat.org/en/latest/guides/traversing-pages.html).
2. Likewise, I modified `iShouldBeLoggedIn`, `clickLogo`, `clickTab`, `clickNavigationZone` to remove the need to inject the custom JavaScript code and to use Mink's API. For the last two method, we need to use `xpath` since we are looking for a label in the page.
3. I removed `activateJsErrorHandler`. Actually, I wonder if we should really test the JavaScript errors here. Most likely, if a JavaScript happens, something else should fail. But if it happens that we want to track the JavaScript error, then this should be done in a more global way and we should provide more details (here basically only the error message is provided).

(Some scenarios are commented just because I did not have time to look at them but some are outdated like the first commented one)

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,423 @@
+     * @param   string  $jsCode
+     * @return  mixed
+     */
+    protected function evalJavascript( $jsCode, $addJsHelper = true )
```
most likely we should be able to get rid of this method and `execJavascript` but one thing though, with `$addJsHelper` set to true (default value), the whole `PlatformUI.js` is read and injected in the page again and again.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,423 @@
+     */
+    protected function waitForLastJs()
+    {
+        //Needs to be here
```
this is very confusing and instead we should be able to read the *why* it needs to be here.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    /**
+     * Waits for Javascript to finnish by running a empty Javascript
```
This is an obvious copy paste error but please pay attention to that.

dpobel commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    /**
+     * @Given I fill in :field subform with:
```
what is a *subform* ? Most likely, the original sentence should be written differently

dpobel commented on Features/DriverJS/Sahi.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public function wrappedFunction( $code )
+    {
+        $jsCode = "(function(){
```
actually, if we ever need to inject some custom JavaScript code, we should always put it inside such a function. So to me, the whole `DriverJSInterface` thing seems useless.

dpobel commented on Features/Lib/PlatformUI.js:
```diff
@@ -0,0 +1,407 @@
+ */
+function PlatformUIHelper(){
+    //YUI global object with node-event-simulate module
+    var Y = YUI().use('node-event-simulate', 'stylesheet');
```
This is wrong, sometimes fails and makes the test unreliable because the YUI module loading is an asynchronous process so it happens that those 2 modules are not loaded while we try to use what they provide.

But in any case, I find weird to rely on YUI in the test and hopefully, this complete file can be removed.

dpobel commented on Features/Lib/PlatformUI.js:
```diff
@@ -0,0 +1,407 @@
+     * @param {String} selector     element selector
+     */
+    this.changeCssDisplay = function( selector ){
+        Y.StyleSheet('NewStyle').set( selector, { display: 'inline' });
```
For the record, you don't to use a dynamically generated stylesheet for that.
`Y.one(selector).setStyle('display', 'inline');` would do exactly the same (in more reliable manner)

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Then I should see "Welcome to eZ Platform" title
+
+    @javascript
+    Scenario: As a admin User, I want to login to PlatformUI
```
isn't it exactly the same as the previous scenario ?

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Then I should see "Activity content" text
+
+    @javascript
+    Scenario: As a admin User, I want to use side menu
```
the name of scenario is bad it seems, it has nothing to do with a *side menu* (side menu has no meaning, I guess you are referring the *action bar* or *discovery bar* but none of those relevant anyway)

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Then I should see "Welcome to eZ Platform" title
+
+    @javascript
+    Scenario: As a admin User, I want to see navigate menu
```
*I want to use the navigation zone menu*

dpobel commented:
One last thing, the way the PR are split makes the review very complicated. A quite big part of the patch is kind of dead code if I don't go to others PR and to be honest this will be very complicated to merge it this way.
